src/metabase/models/serialization/base.clj

```diff
@@ -472,3 +498,136 @@
+  hierarchy, or nil if nothing matches.
+
+  The set of matchers is maintained in an atom, and each model should call [[register-ingestion-path]] to add its
+  matcher.
+
+  Note that the input format is the same as the return value of [[storage-path]]."
+  [path]
+  (first (keep #(% path) (vals @ingest-path-matchers))))
+
+(defn register-ingestion-path
+  "Registers the matcher for the given model. Expects the model name to be a string.
+  The matcher is a function from a path (see [[ingest-path]] for the format) to a `:serdes/meta` hierarchy, or nil
+  if there's no match."
+  [model-name matcher]
+  (swap! ingest-path-matchers assoc model-name matcher)
+  ['register-ingestion-path model-name])
+
+(defn split-leaf-file-name
+  "Given a leaf file name of the type generated by [[storage-leaf-file-name]], break it apart into an [id slug] pair,
+  where the slug might be nil."
+  [file]
+  (when-let [[_ id slug] (or (re-matches #"^([A-Za-z0-9_\.:-]{21})_(.*)$" file)    ; entity_id and slug
+                             (re-matches #"^([A-Za-z0-9_\.:-]{21})$"      file)    ; entity_id only
+                             (re-matches #"^([a-fA-F0-9]{8})_(.*)$"       file)    ; Hash and slug
+                             (re-matches #"^([a-fA-F0-9]{8})$"            file))]  ; Hash only
+    [id slug]))
+
+(defn ingestion-matcher-collected
+  "A helper for the common case of paths like `collections/some/nested/collections/model-name/entityID_slug`.
+  Expects the (lowercase) first segment, and the model name (eg. \"Dashboard\", not \"dashboards\" as it appears in
+  the path).
+  For example `(ingestion-matcher-collected \"collections\" \"Card\")`.
+
+  Returns a matcher function.
+  The resulting hierarchy is "
+  [first-segment model-name]
+  (fn [path]
+    (let [head         (first path)
+          [model file] (take-last 2 path)]
+      (when-let [[id slug] (and (= head first-segment)
+                                (= model (lower-plural model-name))
+                                (split-leaf-file-name file))]
+        (cond-> {:model model-name :id id}
+          slug (assoc :label slug)
+          true vector)))))
+
+(defn- match-pairs [path pattern]
+  (let [chunks (take (count pattern) (partition 2 path))]
+    (when (= pattern (map first chunks))
+      (reduce (fn [out [k v]] (assoc out k v)) {} chunks))))
+
+(defn ingestion-matcher-pairs
+  "A helper for the common case of paths like `databases/my-db/schemas/my-schema/tables/my-table` which alternate
+  fixed and arbitrary segments.
+  The input is a *list* of sequences like `[[\"databases\" \"schemas\" \"tables\"] [\"databases\" \"tables\"]]` and the
+  response is a map of those fixed tokens as keys to the following segment as values, eg.
+  `{\"databases\" \"my-db\"  \"schemas\" \"my-schema\"  \"tables\" \"my-table\"}`.
+  This matches a *prefix*, not necessarily the entire sequence."
+  [path patterns]
+  (first (keep (partial match-pairs path) patterns)))
```

Comment:
Would this be simpler if you used `some` instead of `first` + `keep` ?

Reply:
It certainly would. Done.